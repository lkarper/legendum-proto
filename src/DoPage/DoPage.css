.DoPage__container {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
    height: calc(100vh - 125px);
}

.DoPage__image {
    position: absolute;
    bottom: 25px;
    margin: 0 auto;
    left: 0;
    right: 0;
    max-height: 75%
}

.DoPage__text-container {
    position: absolute;
    top: 50%;
    left: 0;
    min-height: 50%;
    width: 100vw;
    padding: 5px;
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    font-family: 'Crimson Text', serif;
    background: url('/images/paper-background-2-small.png') repeat-y center center fixed;
    -webkit-background-size: cover;
    -moz-background-size: cover;
    -o-background-size: cover;
    background-size: cover;
}

.DoPage__text-container-end {
    position: absolute;
    bottom: 0;
    left: 0;
    height: 50%;
    width: 100vw;
    padding: 5px;
    display: flex;
    flex-direction: column;
    justify-content: space-evenly;
}

.DoPage__background-container {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    margin: 0;
    height: 50%;
}

.DoPage__background-image {
    z-index: -10;    
    height: 100%;
    width: 100%;
    display: block;
}

.DoPage__do-link {
    align-self: center;
}
.DoPage__check-answer-container {
    height: 30px;
}

.DoPage__text {
    margin: 0;
    padding: 5px;
}

.DoPage__button-container {
    width: 100%;
    display: flex;
    flex-direction: row;
    justify-content: space-evenly;
}

.DoPage__nav-button {
    width: 75px;
}

.DoPage__end-quiz-container {
    display: flex;
    flex-direction: column;
    align-content: center;
    align-self: center;
}

.DoPage__h2 {
    text-align: center;
    margin: 5px;
    border-bottom: 1px solid black;
    font-weight: 400;
    font-size: 24px;
}
