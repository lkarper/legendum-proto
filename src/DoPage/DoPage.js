import React, { useState, useContext } from 'react';
import { Link, withRouter } from 'react-router-dom';
import TokenService from '../services/token-service';
import ProgressService from '../services/progress-service';
import UserContext from '../contexts/UserContext';
import UserIncorrect from '../UserIncorrect/UserIncorrect';
import UserCorrect from '../UserCorrect/UserCorrect';
import QuestionToDisplay from '../QuestionToDisplay/QuestionToDisplay';

const DoPage = (props) => {

    const context = useContext(UserContext);

    const [userCorrect, setUserCorrect] = useState();
    const [userIncorrect, setUserIncorrect] = useState();
    const [error, setError] = useState(null);

    const { chapt, savedUserInput, pages, page, setPage, setSavedUserInput } = props.data;
    const pageToDisplay = pages[page - 1];

    const checkAnswer = (event, userResponse) => {
        event.preventDefault();
        if (pageToDisplay.look_ahead) {
            setSavedUserInput({
                ...savedUserInput,
                [pageToDisplay.property_to_save]: userResponse,
            });
            setUserIncorrect();
            setUserCorrect(userResponse);
        } else if (userResponse === pageToDisplay.correct_response) {
            setUserIncorrect();
            setUserCorrect(userResponse);
        } else {
            setUserCorrect();
            setUserIncorrect(userResponse);
        }
    }

    const onCompletion = () => {
        if (TokenService.hasAuthToken()) {
            ProgressService.postProgress(chapt)
                .then(progressObject => {
                    setError(null);
                    context.updateProgress(progressObject);
                    props.history.push(`/game/story/${parseInt(chapt) + 1}`);
                })
                .catch(error => {
                    setError(error.message);
                });
        }
    }

    return (
        <div>
            <img 
                src={pageToDisplay.image_url}
                alt={pageToDisplay.image_alt_text}
            />
            <QuestionToDisplay 
                pageToDisplay={pageToDisplay}
                savedUserInput={savedUserInput}
                checkAnswer={checkAnswer}
            />
            {userIncorrect ? <UserIncorrect page={pageToDisplay} userResponse={userIncorrect} /> : ''}
            {userCorrect ? <UserCorrect page={pageToDisplay} userResponse={userCorrect} /> : ''}
            {page !== 1 ? 
                <button 
                    onClick={() => {
                        setUserCorrect();
                        setUserIncorrect();
                        setPage(page - 1);
                    }}
                >
                    &#60;
                </button> 
                : ''
            }
            {page !== pages.length && userCorrect ? 
                <button 
                    onClick={() => {
                        setUserCorrect();
                        setUserIncorrect();
                        setPage(page + 1);
                    }}
                >
                    &#62;
                </button> 
                : ''
            }
            <div role="alert">
                {page === pages.length && userCorrect
                    ? 
                        <>
                            {TokenService.hasAuthToken() 
                                ? <> 
                                    <button
                                        onClick={onCompletion} 
                                    >
                                        On to the next chapter.
                                    </button>
                                    <Link
                                        to='/dashboard'
                                    >Back to dashboard (progress will not be saved)</Link>
                                </>
                                : <Link to={`/game/story/${parseInt(chapt) + 1}`}>On the the next chapter</Link>
                            }
                            {/* Where else should the user be able to go? Some sort of home? A leaderboard? A group? */}
                        </> 
                    : ''
                }
                {error ? 
                    <>
                        <h2>Could not save progress: {error} </h2>
                        <p>Click below to try again (you will be redirected to the next chapter if successful).</p>
                        <button onClick={onCompletion}>Try again</button>
                    </>
                    : ''
                }
            </div>
        </div>
    );
}

export default withRouter(DoPage);