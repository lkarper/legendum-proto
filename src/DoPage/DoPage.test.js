import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import DoPage from './DoPage';

describe('DoPage component', () => {
  it('renders the UI without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <DoPage />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders the UI as expected', () => {
    const wrapper = mount(
      <BrowserRouter>
        <DoPage />
      </BrowserRouter>
    );
    expect(toJSON(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
