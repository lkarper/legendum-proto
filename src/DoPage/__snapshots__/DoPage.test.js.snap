// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DoPage component renders the UI as expected 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(DoPage)>
      <DoPage
        chapt=""
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "createHref": [Function],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        match={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        page={1}
        pages={Array []}
        savedUserInput={Object {}}
        setPage={[Function]}
        setSavedUserInput={[Function]}
      >
        <p>
          Error: Looks like something went wrong. Check your connection and try again.
        </p>
      </DoPage>
    </withRouter(DoPage)>
  </Router>
</BrowserRouter>
`;
